### PR TITLE
Events V2: Create events

### PR DESCRIPTION
**Pull request by rikukissa**

### CODE REVIEW

File: packages/client/src/v2-events/features/events/useEvents.ts [hunk 1]
```diff
@@ -10,76 +10,138 @@
...
+  // the action will be retried once
+  retry: 1,
+  retryDelay: 5000,
+  mutationFn: async (actionInput) => {
+    /*
+     * Handles action being called with an outdated temporary id
+     *
+     * If you have created the event offline and you've also created the action
+     * offline, then once you come back online it might just happen that you have
+     * an action request in the buffer that still uses the temporary id to refer to the event
```
Comment (makelicious) [inline packages/client/src/v2-events/features/events/useEvents.ts:32]: Could be its own function. jsdoc would be targeted nicely

File: packages/client/src/v2-events/features/events/useEvents.ts [hunk 2]
```diff
@@ -10,76 +10,138 @@
...
+import { api, queryClient, utils } from '@client/v2-events/trpc'
 import { EventDocument } from '@events/schema'
-import { useQueryClient, useSuspenseQuery } from '@tanstack/react-query'
+import { useSuspenseQuery } from '@tanstack/react-query'
 
 export function preloadData() {
   utils.config.get.ensureData()
 }
 
+utils.event.actions.declare.setMutationDefaults(({ canonicalMutationFn }) => ({
```
Comment (makelicious) [inline packages/client/src/v2-events/features/events/useEvents.ts:21]: Do you have any issues when running this inline?

I receive something along the lines of "cannot use utils until it has been initialized"

File: packages/client/src/v2-events/features/events/useEvents.ts [hunk 3]
```diff
@@ -10,76 +10,138 @@
...
 utils.event.create.setMutationDefaults(({ canonicalMutationFn }) => ({
-  mutationFn: canonicalMutationFn
+  mutationFn: canonicalMutationFn,
+  retry: 0,
+  onMutate: async (newEvent) => {
+    const optimisticEvent = {
+      id: newEvent.transactionId,
+      type: newEvent.type,
+      transactionId: newEvent.transactionId,
+      createdAt: new Date(),
```
Comment (makelicious) [inline packages/client/src/v2-events/features/events/useEvents.ts:91]: Should we opt for toISOString()  with dates? easier to manage types end to end.

File: packages/client/src/v2-events/features/events/useEvents.ts [hunk 4]
```diff
@@ -10,76 +10,138 @@
...
+      updatedAt: new Date(),
+      actions: []
+    }
+    console.log('setting it now', optimisticEvent)
+    queryClient.cancelQueries({ queryKey: ['persisted-events'] })
```
Comment (makelicious) [inline packages/client/src/v2-events/features/events/useEvents.ts:96]: 'persisted-events' could be constant

File: packages/client/src/v2-events/features/events/useEvents.ts [hunk 5]
```diff
@@ -10,76 +10,138 @@
...
   const data = await storage
-    .getItem('events')
-    .then((e) => (e ? JSON.parse(e) : []))
+    .getItem<EventDocument[]>('events')
+    .then((e) => e || [])
 
   return data
 }
 
 function writeEventToStorage(events: EventDocument[]) {
```
Comment (makelicious) [inline packages/client/src/v2-events/features/events/useEvents.ts:135]: `writeEventsToStorage`

**Review by makelicious — APPROVED** [review]

File: packages/client/src/v2-events/features/events/useEvents.ts (hunk 2 above)
Comment (rikukissa) [reply packages/client/src/v2-events/features/events/useEvents.ts:21]: Noticed this as well but it got cleared away automatically. Something about import/export order and evaluation. Not yet 100% sure

File: packages/client/src/v2-events/features/events/useEvents.ts (hunk 3 above)
Comment (rikukissa) [reply packages/client/src/v2-events/features/events/useEvents.ts:110]: If I understand correctly, `superjson` transforms the Date to ISO string before sending and then transforms it back when backend receives it

File: packages/client/src/v2-events/features/events/EventFormWizard.tsx [hunk 6]
```diff
@@ -9,106 +9,218 @@
...
+  }
+})
+
+function EventFormWizard() {
+  const { eventId } = useParams<{
+    eventId: string
+  }>()
+  const events = useEvents()
+
+  const [event] = events.getEvent(eventId)
```
Comment (naftis) [inline packages/client/src/v2-events/features/events/EventFormWizard.tsx:89]: You may want to consider creating `useEvent(eventId)` hook

File: packages/components/src/FormWizard/FormWizard.tsx [hunk 7]
```diff
@@ -8,81 +8,62 @@
...
-          ) : (
-            // Initial simple submit for testing
-            <Button type="primary">Submit</Button>
-          )}
-        </Stack>
-      </form>
-    </FormProvider>
+            {currentPage + 1 < totalPages ? (
+              <Button type="primary" role="button" onClick={onNextPage}>
+                Continue
```
Comment (naftis) [inline packages/components/src/FormWizard/FormWizard.tsx:57]: just remember to i18n these

File: packages/client/src/v2-events/features/events/useEvents/useEvents.ts [hunk 8]
```diff
@@ -0,0 +1,289 @@
...
+}
+
+/**
+ * Wraps a canonical mutation function to handle outdated temporary IDs.
+ *
+ * When an event is created offline and actions referring to that event
+ * are also created offline, there may be cases where a request in the
+ * buffer still uses a temporary ID to reference the event. This wrapper
+ * ensures that the `eventId` parameter is updated to its canonical value
+ * before the mutation function is called.
```
Comment (naftis) [inline packages/client/src/v2-events/features/events/useEvents/useEvents.ts:47]: would allowing arbitrary uuid for the eventId make this simpler?

**Review by naftis — COMMENTED** [review]
👍 👍 

a thing to consider is how do we keep the middleware in `useEvents` readable in the future as well

**Pull request merged**